docs/tech-specs/basic-spec.md

```diff
@@ -6,6 +6,269 @@
+
+### Storage Choice
+
+Storage would be the most obvious bottleneck in scaling throughput for an application such as this. But I firmly believe too much concern of such cases is premature optimization and it would be nice to have this headache. So we will go for simplicity - MySQL 8. Primarily because if push comes to shove we will be able to choose [MySQL Group Replication](https://dev.mysql.com/doc/refman/8.0/en/group-replication.html) with [multi-primary mode enabled](https://dev.mysql.com/doc/refman/8.0/en/group-replication-multi-primary-mode.html); this functionality is also available through [managed services](https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/aurora-multi-master.html#aurora-multi-master-overview). The implementation will be abstracted sufficiently we can totally switch to a NoSQL such as Kafka, Cassandra for scaling writes. For unit tests we will be using SQLite. Furthermore, GoLang's [MySQL driver](https://github.com/go-sql-driver/mysql) is matured from some [extensive production use](https://github.blog/2020-05-20-three-bugs-in-the-go-mysql-driver/) as well.
+
+### Prioritized Message Limitations
+
+Taking priority into consideration automatically will have throughput implications since enqueueing and dequeueing will have to be mutually exclusive across the queue. And trying to maintain the queue across the cluster of broker would add more complication. So we will limit the priority to take effect within a single instance of the broker process. What it means is when we have a worker available to dispatch a message it will always give priority to the highest priority message. It also means, if there are too many messages queued, the likely time needed to queue will be longer; so increasing brokers aggressively is a deployment recommendation.
+
+### HTTP2 Consideration
+
+Given throughput is one of the primary goal of the project, HTTP/2 had to be an avenue we explore to take advantage of its multiplexing capabilities and binary frame compression. The limitation would be what if a consumer is not HTTP/2 enabled or the producer's HTTP client is HTTP/1.1 compliant only; in most cases Go's HTTP server and client can talk in both protocols. One other consideration is, the load balancer used with the broker has to be HTTP/2 compliant in receiving and proxying; for example, until [very recently](https://aws.amazon.com/blogs/aws/new-application-load-balancer-support-for-end-to-end-http-2-and-grpc/) AWS ALB is HTTP/2 enabled in receiving but not when proxying. From an implementation standpoint we will have to account for the fact that now requests [will come more in batch](https://www.lucidchart.com/techblog/2019/04/10/why-turning-on-http2-was-a-mistake/) rather than sequential calls.
+
+## Implementation Details
+
+### Objects
+
+![Class Diagram](./images/basic-spec/BrokerObjects.svg)
```

Comment:
why does data on producers need to be stored? wouldn't it be more simple for any process to send messages to a channel without having to register as a producer?

Reply:
Good question; it is to add some level of traceability of where a message came from; specifically which system. Without producer it would be difficult to have a reasonable trace. Furthermore, later if we want to introduce some kind of control mechanism of "who" can broadcast producer will come in handy. Currently we do not want "anyone" to broadcast, we want "someone" to broadcast.